executor/write.go

```diff
@@ -907,69 +928,260 @@ func getKeysNeedCheck(ctx sessionctx.Context, t table.Table, rows [][]types.Datu
+			if err1 != nil {
+				return nil, errors.Trace(err1)
+			}
+			keysWithErr = append(keysWithErr, keyWithDupError{
+				false,
+				key,
+				kv.ErrKeyExists.FastGen("Duplicate entry '%s' for key '%s'",
+					colValStr, v.Meta().Name),
+				newRowValue,
+			})
 		}
 		rowWithKeys = append(rowWithKeys, keysWithErr)
 	}
 	return rowWithKeys, nil
 }
 
-// batchMarkDupRows marks rows with duplicate errors as nil.
-// All duplicate rows were marked and appended as duplicate warnings
-// to the statement context in batch.
-func batchMarkDupRows(ctx sessionctx.Context, t table.Table, rows [][]types.Datum) ([][]types.Datum, error) {
-	// get keys need to be checked
-	rowWithKeys, err := getKeysNeedCheck(ctx, t, rows)
+// batchGetInsertKeys uses batch-get to fetch all key-value pairs to be checked for ignore or duplicate key update.
+func batchGetInsertKeys(ctx sessionctx.Context, t table.Table, newRows [][]types.Datum) ([][]keyWithDupError, map[string][]byte, error) {
+	// Get keys need to be checked.
+	keysInRows, err := getKeysNeedCheck(ctx, t, newRows)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
 
-	// batch get values
+	// Batch get values.
 	nKeys := 0
-	for _, v := range rowWithKeys {
+	for _, v := range keysInRows {
 		nKeys += len(v)
 	}
 	batchKeys := make([]kv.Key, 0, nKeys)
-	for _, v := range rowWithKeys {
+	for _, v := range keysInRows {
 		for _, k := range v {
 			batchKeys = append(batchKeys, k.key)
 		}
 	}
 	values, err := kv.BatchGetValues(ctx.Txn(), batchKeys)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	return keysInRows, values, nil
+}
+
+// checkBatchLimit check the batchSize limitation.
+func (e *InsertExec) checkBatchLimit() error {
+	sessVars := e.ctx.GetSessionVars()
+	batchInsert := sessVars.BatchInsert && !sessVars.InTxn()
+	batchSize := sessVars.DMLBatchSize
+	if batchInsert && e.rowCount >= batchSize {
+		e.ctx.StmtCommit()
+		if err := e.ctx.NewTxn(); err != nil {
+			// We should return a special error for batch insert.
+			return ErrBatchInsertFail.Gen("BatchInsert failed with error: %v", err)
+		}
+		e.rowCount = 0
+		if !sessVars.ImportingData {
+			sessVars.GetWriteStmtBufs().BufStore = kv.NewBufferStore(e.ctx.Txn(), kv.TempTxnMemBufCap)
+		}
+	}
+	return nil
+}
+
+// initDupOldRowValue initializes dupOldRowValues which contain the to-be-updated rows from storage.
+func (e *InsertExec) initDupOldRowValue(newRows [][]types.Datum) (err error) {
+	e.dupOldRowValues = make(map[string][]byte, len(newRows))
+	handles := make([]int64, 0, len(newRows))
+	for _, keysInRow := range e.uniqueKeysInRows {
+		for _, k := range keysInRow {
+			if val, found := e.dupKeyValues[string(k.key)]; found {
+				var handle int64
+				handle, err = decodeOldHandle(k, val)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				handles = append(handles, handle)
+				break
+			}
+		}
+	}
+	e.dupOldRowValues, err = batchGetOldValues(e.ctx, e.Table, handles)
```

Comment:
if k.isRecordKey is true,  unnecessary to fetch values again, use value in e.dupKeyValues.